internal/client/consensus/grandpa/communication_gossip.go

```diff
@@ -0,0 +1,1708 @@
+	defer gv.innerMtx.Unlock()
+	gv.inner.peers.peerDisconnected(who)
+}
+
+func (gv *gossipValidator[H, N, Hasher]) Validate(
+	context gossip.ValidatorContext[H],
+	who peerid.PeerID,
+	data []byte,
+) gossip.ValidationResult {
+	action, broadcastTopics, peerReply := gv.doValidate(who, data)
+
+	// not with lock held!
+	if peerReply != nil {
+		var gossipMessage gossipMessageVDT[H, N]
+		gossipMessage.inner = peerReply
+		context.SendMessage(who, scale.MustMarshal(gossipMessage))
+	}
+
+	for _, topic := range broadcastTopics {
+		context.SendTopic(who, topic, false)
+	}
+
+	switch action := action.(type) {
+	case actionKeep[H]:
+		gv.report(who, action.ReputationChange)
+		context.BroadcastMessage(action.Hash, data, false)
+		return gossip.ValidationResultProcessAndKeep[H]{Hash: action.Hash}
+	case actionProcessAndDiscard[H]:
+		gv.report(who, action.ReputationChange)
+		return gossip.ValidationResultProcessAndDiscard[H]{Hash: action.Hash}
+	case actionDiscard[H]:
+		gv.report(who, action.ReputationChange)
+		return gossip.ValidationResultDiscard{}
+	default:
+		panic("unreachable")
+	}
+}
+
+func (gv *gossipValidator[H, N, Hasher]) MessageAllowed() func(
+	who peerid.PeerID,
+	intent gossip.MessageIntent,
+	topic H,
+	data []byte,
+) bool {
+	var doRebroadcast bool
+	{
+		gv.innerMtx.Lock()
+		defer gv.innerMtx.Unlock()
+		now := time.Now()
+		if now.After(gv.inner.nextRebroadcast) || now.Equal(gv.inner.nextRebroadcast) {
+			gv.inner.nextRebroadcast = now.Add(rebroadcastAfter)
+			doRebroadcast = true
+		} else {
+			doRebroadcast = false
+		}
+	}
+
+	return func(who peerid.PeerID, intent gossip.MessageIntent, topic H, data []byte) bool {
+		gv.innerMtx.RLock()
+		defer gv.innerMtx.RUnlock()
+
+		if intent == gossip.MessageIntentPeriodicRebroadcast {
+			return doRebroadcast
+		}
+
+		peer := gv.inner.peers.peer(who)
+		if peer == nil {
+			return false
+		}
+
+		// if the topic is not something we're keeping at the moment,do not send.
+		entry := gv.inner.liveTopics.topicInfo(topic)
+		if entry == nil {
+			return false
+		}
+		maybeRound, setID := entry.Round, entry.SetID
+
+		if intent == gossip.MessageIntentBroadcast {
+			if maybeRound != nil {
+				// early return if the global message isn't allowed at this stage.
+				if !gv.inner.roundMessageAllowed(who) {
+					return false
+				}
+			} else if !gv.inner.globalMessageAllowed(who) {
+				// early return if the global message isn't allowed at this stage.
+				return false
+			}
+		}
+
+		// if the topic is not something the peer accepts, discard.
+		if maybeRound != nil {
+			return peer.view.considerVote(*maybeRound, setID) == considerAccept
+		}
+
+		// global message.
+		if gv.inner.localView == nil {
+			// cannot evaluate until we have a local view.
+			return false
+		}
+		var gossipMessage gossipMessageVDT[H, N]
+		err := scale.Unmarshal(data, &gossipMessage)
+		if err != nil {
+			return false
+		}
+		val, err := gossipMessage.Value()
+		if err != nil {
+			return false
+		}
```

Comment:
can we log such errors? here we obfuscate them which might be a nightmare to debug/investigate any problem that occurs here.